src/mame/drivers/brother_lw30.cpp

```diff
@@ -0,0 +1,1144 @@
+// license:BSD-3-Clause
+// copyright-holders:Bartman/Abyss
+
+#include "emu.h"
+#include "emupal.h"
+#include "screen.h"
+#include "speaker.h"
+#include "machine/timer.h"
+#include "cpu/z180/z180.h"
+#include "debug/debugcpu.h"
+#include "sound/beep.h"
+#include "video/mc6845.h"
+
+// if updating project, c:\msys64\win32env.bat
+// cd \schreibmaschine\mame_src
+// make SUBTARGET=schreibmaschine NO_USE_MIDI=1 NO_USE_PORTAUDIO=1 vs2019
+
+// command line parameters:
+// -log -debug -window -intscalex 2 -intscaley 2 lw30 -resolution 960x256 -flop roms\lw30\tetris.img
+// -debug -autoboot_script c:\schreibmaschine\mame_src\lw30.lua -log -window -intscalex 8 -intscaley 8 -resolution 1920x512 lw30 -flop c:\brothers\mame\disks\lw30\tetris.img 
+
+#pragma region(LW-30)
+
+//////////////////////////////////////////////////////////////////////////
+// LW-30
+//////////////////////////////////////////////////////////////////////////
+
+/***************************************************************************
+
+Brother LW-30
+1991
+
+Hardware:
+
+#7 
+Hitachi
+HD64180RP6
+8-bit CMOS Micro Processing Unit
+fully compatible with Zilog Z80180 (Z180)
+6 MHz, DP-64S, Address Space 512 K Byte
+MuRata CST12MTW 12.00 MHz Ceramic Resonator
+
+#8 
+Mitsubishi
+M65122ASP
+UA5445-B LC-1
+
+#6 
+NEC
+D23C4001EC-172
+UA2849-A
+4MBit Mask ROM for Dictionary
+
+#5
+LH532H07
+UA5362-A
+2MBit Mask ROM
+
+#11 
+Hitachi
+HM6264ALP-15
+High Speed CMOS Static RAM (8kbit x 8) 150ns
+
+#10
+Mitsubishi
+M65017FP
+UA5498-A
+Murata CST4MGW 4.00 MHz Ceramic Resonator
+
+#3, #4
+Mitsubishi
+HD74LS157P
+
+#1, #2
+NEC
+D41464C-10
+Dynamic NMOS RAM (64kbit x 4) 100ns
+
+QA1, QA2
+Mitsubishi
+M54587P
+
+#12
+Texas Instruments
+SN74HC04N
+
+Floppy - custom single sided 3.5" DD
+240kb capacity
+256 bytes/sector
+12 sectors/track
+78 tracks
+custom 5-to-8 GCR encoding (very similar to Apple II's 5-and-3 encoding)
+300 rpm
+FF FF FF used as sync-start, AB sync-mark for sector header, DE sync-mark for sector data
+FAT12 File System
+
+ROHM
+BA6580DK
+Read/Write Amplifier for FDD
+
+Emulation Status:
+Printer not working
+Floppy Disk writing not working
+Floppy Disk Sync not implemented (reading works)
+Dictionary ROM not working
+Cursor shapes not implemented except block cursor
+
+TODO: find self-test; verify RAM address map
+// 8kb SRAM, 64kb DRAM <- where?
+
+***************************************************************************/
+
+#pragma region(Floppy)
+
+class lw30_floppy_image_device;
+
+class lw30_floppy_connector : public device_t, public device_slot_interface
+{
+public:
+	lw30_floppy_connector(const machine_config &mconfig, const char *tag, device_t *owner, uint32_t clock);
+
+	lw30_floppy_image_device *get_device();
+
+protected:
+	void device_start() override {}
+	void device_config_complete() override {}
+};
+
+class lw30_floppy_image_device : public device_t, public device_image_interface
```

Comment:
Why is this not implemented as a floppy format?

Reply:
probably because I don't know how to do that. This is a custom GCR encoding. Would that work?

Reply:
Yes, what details do you have about this encoding?

Reply:
see lines 179-336 for my implementation. Any pointers on how to convert that code to a floppy format?

Reply:
I have now implemented it as a floppy format.